analyzer/src/steps/resolve.rs

```diff
@@ -248,36 +310,27 @@ impl<'a, 'e> SymbolResolver<'a, 'e> {
                                     .unwrap();
 
                                 let import_symbol = ResolvedSymbol::new(env_id, var_id);
-                                resolved_imports.set_import(var.name.clone(), import_symbol);
+                                imports.set_resolved_import(
+                                    var.name.clone(),
+                                    ResolvedImport::Symbol(import_symbol),
+                                );
                             }
                         }
                     }
                 }
             }
         }
-        resolved_imports
     }
 
-    /// Gets an environment from the given fully qualified name, then pop the name until it finds a valid environment.
-    /// returns None if the name's root could not get resolved
-    fn get_env_from(&self, name: &Name) -> Option<(SourceObjectId, &'a Environment)> {
-        let mut env_name = Some(name.clone());
-        while let Some(name) = env_name {
-            if let Some((id, env)) = self.engine.find_environment_by_name(&name) {
-                return Some((id, env));
-            }
-            env_name = name.tail();
-        }
-        None
-    }
+    /// Iterates over remaining unresolved symbols
+    /// This resolution supports parent lookups if the resolution could not
+    fn resolve_trees(&mut self, to_visit: &mut Vec<Name>, visited: &mut HashSet<Name>) {
+        //NOTE: the resolution does not insert new objects
+        'symbol: for object_id in 0..self.relations.objects.len() {
+            let object = &self.relations.objects[object_id];
+            let object_id = GlobalObjectId(object_id);
```

Comment:
```suggestion
        'symbol: for (object_id, object) in self.relations.objects.iter_mut().enumerate() {
            let object_id = GlobalObjectId(object_id);
```
Maybe move `imports` outside `Resolver` to be able to call `relations.iter_mut()` directly

Reply:
the borrowing issue here is because the resolver is mutably borrowed inside the loop when object states are modified